ansible_runner/streaming.py

```diff
@@ -81,6 +89,64 @@ def __init__(self, _input=None, _output=None, **kwargs):
         self.status = "unstarted"
         self.rc = None
 
+    def _begin_keepalive(self):
+        """Starts a keepalive thread at most once"""
+        if not self._keepalive_thread:
+            self._keepalive_thread = Thread(target=self._keepalive_loop)
+            self._keepalive_thread.start()
+
+    def _end_keepalive(self):
+        """Disable the keepalive interval and notify the keepalive thread to shut down"""
+        self._keepalive_interval_sec = 0
+        self._output_event.set()
+
+    def _keepalive_loop(self):
+        """Main loop for keepalive injection thread; exits when keepalive interval is <= 0"""
+        while self._keepalive_interval_sec > 0:
+            # block until output has occurred or keepalive interval elapses
+            if self._output_event.wait(timeout=self._keepalive_interval_sec):
+                # output was sent before keepalive timeout; reset the event and start waiting again
+                self._output_event.clear()
+                continue
+
+            # keepalive interval elapsed; try to send a keepalive...
+            # pre-acquire the output lock without blocking
+            if not self._output_lock.acquire(blocking=False):
+                # something else has the lock; output is imminent, so just skip this keepalive
+                # NB: a long-running operation under an event handler that's holding this lock but not actually moving
+                # output could theoretically block keepalives long enough to cause problems, but it's probably not
+                # worth the added locking hassle to be pedantic about it
+                continue
+
+            try:
+                # were keepalives recently disabled?
+                if self._keepalive_interval_sec <= 0:
+                    # we're probably shutting down; don't risk corrupting output by writing now, just bail out
+                    return
+                # output a keepalive event
+                # FIXME: this could be a lot smaller (even just `{}`) if a short-circuit discard was guaranteed in
+                #  Processor or if other layers were more defensive about missing event keys and/or unknown dictionary
+                #  values...
+                self.event_handler(dict(event='keepalive', counter=0, uuid=0))
```

Comment:
I would prefer to lose both of `counter=0, uuid=0`, since neither value makes sense or adds any value. I absolutely want to keep it as a single-line JSON, but the eof event is really just `{"eof": true}`

Reply:
I'd love to make it super tiny (even `{}`), but I don't know at what point it's safe to assume that an old broken `Process` impl won't encounter this and barf on it- if the overall keepalive feature doesn't end up being backported, maybe that's a reasonable approach, but if it is, we'll need to have something (at least for a while) that won't break an older processor implementation.

Reply:
@shanemcd and I were talking about how to introduce this as a hotfix to an existing environment via an envvar that could be set in the EE itself, rather than having to plumb it all the way through as a kwarg to `Worker` and (somehow) deal with older EEs that don't understand that kwarg (possibly even just enabling a sane default in *all* our EEs and calling it a day). This is another one of those cases where the payload format design and its consumers didn't consider upgrade-ability, so either we force everything to be upgraded, or we stick with a format that's non-objectionable to existing consumers.

Reply:
* added `ANSIBLE_RUNNER_KEEPALIVE_SECONDS` envvar fallback; CLI/kwarg value takes precedence
* cleaned up the transmit kwargs so we'll never send `keepalive_seconds` in a payload kwargs bundle (as it's nonsensical there and old runners will barf on it)
* fixed `Process` to short-circuit drop keepalives and drop any malformed events (currently silently; if there's a good place to do so loudly, I'm all ears)

Still need to decide if we want to allow a minimal hotfix deployment for this with only the worker EE- if not, we can truncate the extra crap off the keepalive payload. Another option would be to only do the fatter payload on the 2.3-equivalent version of runner and use the smaller payload for anything new (even maybe `{}`)...

Reply:
If it is a goal to have an EE emit keepalive messages in a way that won't fail the API in some way, then from brief testing with AWX, it seems that the minimal requirements would be that `counter` is greater than 0. Then, that the `event` is a value within the listed event types here:

https://github.com/ansible/awx/blob/0815f935ca359537c023efadf286671e31dc3374/awx/main/models/events.py#L164

And then, to make the UI not mess up, you would ideally want to set start_line and end_line to the same number. What number? I'm not sure.

Reply:
(consensus was "fail fast", so the keepalive doesn't look like a valid "event" in case it escapes from `Process`, it'll just fail the process altogether)